File: 2023/Leetcode/problem104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    // n: number of nodes
    // Time Complexity: O(n) Linear Time
    // Space Complexity: O(n) Linear Space

    // Depth-First Search (DFS) w/recursion
    // Breadth-First Search (BFS) w/Queue
    if (!root) {
        return 0;
    }

    let left = maxDepth(root.left);
    let right = maxDepth(root.right);

    if (left > right) {
        return left + 1;
    }
    return right + 1;
};

File: 2022/Leetcode/problem104.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int deepestDepth = 0;
    
public:
    void findDepth(TreeNode* current, int currentDepth = 1) {
        if (current) {
            if (deepestDepth < currentDepth) {
                deepestDepth = currentDepth;
            }
            
            findDepth(current -> left, currentDepth + 1);
            findDepth(current -> right, currentDepth + 1);
        }
    }
    
    int maxDepth(TreeNode* root) {
        findDepth(root);
        return deepestDepth;
    }
};